### ISSUE TITLE
Work around SQL Server's 4000 character JSON_VALUE length limitation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14463]
SQL string lengths are the gift that keep giving. In SQL Server, `JSON_VALUE` can extract values of at most 4000 characters string length, see:
https://stackoverflow.com/q/45024937/521799

**Example**:

```sql
select 
  len(json_value(j, '$.short')),
  len(json_value(j, '$.long'))
from (
  select (
    select  
      replace(space(4000), ' ', 'x') short, 
      replace(space(4001), ' ', 'x') long
    for json path, without_array_wrapper
  ) j
) t
```

The result is underwhelming (in lax mode, there's an error in strict mode):

```
|     |   |
|-----|---|
|4,000|   |
```

**Workaround:**

```sql
select 
  len((select [value] from openjson(j) where [key] = 'short')),
  len((select [value] from openjson(j) where [key] = 'long'))
from (
  select (
    select  
      replace(space(4000), ' ', 'x') short, 
      replace(space(4001), ' ', 'x') long
    for json path, without_array_wrapper
  ) j
) t
```

Producing

```
|     |     |
|-----|-----|
|4,000|4,001|
```

The workaround isn't simple to apply in this case, because we'd have to translate the JSON path to something equivalent in SQL, which is very hard. Perhaps, better use `OPENJSON()`'s `WITH` clause, see the docs:
https://learn.microsoft.com/en-us/sql/t-sql/functions/openjson-transact-sql?view=sql-server-ver16

**Better workaround**:

```sql
select 
  len((select c from openjson(j) with (c varchar(max) '$.short'))),
  len((select c from openjson(j) with (c varchar(max) '$.long')))
from (
  select (
    select  
      replace(space(4000), ' ', 'x') short, 
      replace(space(4001), ' ', 'x') long
    for json path, without_array_wrapper
  ) j
) t
```

Producing:

```
|     |     |
|-----|-----|
|4,000|4,001|
```

**Comment by lukaseder** [comment]
The emulation only works if the `path` is a `Param`, in case of which we have to extract its value. If it's any other expression, we have to keep the `JSON_VALUE` logic, as `OPENJSON()` doesn't support dynamic paths.

Also, ideally, if we *know* the column is shorter than 4000 characters, we should work with the existing implementation. We can mainly know this if the `Field` is a `TableField` of type `VARCHAR` with length <= 4000

**Comment by lukaseder** [comment]
This might be a change:

1. In terms of performance
2. In terms of subtle change of behaviour

So, I'm not backporting the fix.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0
